File: api/src/models/Community.model.js
const mongoose = require("mongoose");

const communitySchema = new mongoose.Schema(
  {
    name: {
      type: String,
      required: [true, "El nombre de la comunidad es requerido"],
      trim: true,
      minlength: [2, "El nombre debe tener al menos 2 caracteres"],
      maxlength: [100, "El nombre no puede exceder 100 caracteres"],
    },
    address: {
      type: String,
      required: [true, "La dirección es requerida"],
      trim: true,
      minlength: [5, "La dirección debe tener al menos 5 caracteres"],
      maxlength: [200, "La dirección no puede exceder 200 caracteres"],
    },
    description: {
      type: String,
      trim: true,
      maxlength: [500, "La descripción no puede exceder 500 caracteres"],
      default: "",
    },
    monthlyFee: {
      type: Number,
      required: [true, "La mensualidad es requerida"],
      min: [0, "La mensualidad debe ser un número positivo"],
    },
    currency: {
      type: String,
      required: [true, "La moneda es requerida"],
      enum: {
        values: ["MXN", "USD", "EUR"],
        message: "Moneda no válida. Debe ser MXN, USD o EUR",
      },
      default: "MXN",
    },
    totalHouses: {
      type: Number,
      required: [true, "El total de casas es requerido"],
      min: [1, "Debe haber al menos 1 casa"],
      max: [1000, "No puede haber más de 1000 casas"],
    },
    superAdminId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: [true, "El super administrador es requerido"],
    },
    adminIds: [
      {
        type: mongoose.Schema.Types.ObjectId,
        ref: "User",
      },
    ],
    isActive: {
      type: Boolean,
      default: true,
    },
    settings: {
      allowResidents: {
        type: Boolean,
        default: true,
      },
      requireApproval: {
        type: Boolean,
        default: true,
      },
      maxFileSize: {
        type: Number,
        default: 5 * 1024 * 1024, // 5MB en bytes
        min: [1024 * 1024, "El tamaño máximo de archivo debe ser al menos 1MB"],
        max: [
          50 * 1024 * 1024,
          "El tamaño máximo de archivo no puede exceder 50MB",
        ],
      },
      notificationPreferences: {
        email: {
          type: Boolean,
          default: true,
        },
        push: {
          type: Boolean,
          default: true,
        },
        sms: {
          type: Boolean,
          default: false,
        },
      },
    },
    createdAt: {
      type: Date,
      default: Date.now,
    },
    updatedAt: {
      type: Date,
      default: Date.now,
    },
  },
  {
    timestamps: true,
    toJSON: { virtuals: true },
    toObject: { virtuals: true },
  }
);

// Índices para mejorar el rendimiento de las consultas
communitySchema.index({ name: 1 });
communitySchema.index({ superAdminId: 1 });
communitySchema.index({ adminIds: 1 });
communitySchema.index({ isActive: 1 });
communitySchema.index({ createdAt: -1 });

// Virtual para obtener el total de administradores
communitySchema.virtual("totalAdmins").get(function () {
  return this.adminIds && Array.isArray(this.adminIds)
    ? this.adminIds.length
    : 0;
});

// Virtual para obtener el total de residentes (se calculará dinámicamente)
communitySchema.virtual("totalResidents").get(function () {
  // Este valor se calculará dinámicamente desde el modelo User
  return 0;
});

// Middleware para actualizar updatedAt antes de guardar
communitySchema.pre("save", function (next) {
  this.updatedAt = new Date();
  next();
});

// Método estático para buscar comunidades por super admin
communitySchema.statics.findBySuperAdmin = function (superAdminId) {
  return this.find({ superAdminId });
};

// Método estático para buscar comunidades activas
communitySchema.statics.findActive = function () {
  return this.find({ isActive: true });
};

// Método de instancia para agregar administrador
communitySchema.methods.addAdmin = function (adminId) {
  if (!this.adminIds.includes(adminId)) {
    this.adminIds.push(adminId);
    return this.save();
  }
  return Promise.resolve(this);
};

// Método de instancia para remover administrador
communitySchema.methods.removeAdmin = function (adminId) {
  this.adminIds = this.adminIds.filter((id) => !id.equals(adminId));
  return this.save();
};

// Método de instancia para verificar si un usuario es administrador
communitySchema.methods.isAdmin = function (userId) {
  return this.adminIds.some((id) => id.equals(userId));
};

// Método de instancia para verificar si un usuario es super admin
communitySchema.methods.isSuperAdmin = function (userId) {
  return this.superAdminId.equals(userId);
};

// Método de instancia para obtener información pública
communitySchema.methods.getPublicInfo = function () {
  return {
    id: this._id,
    name: this.name,
    address: this.address,
    description: this.description,
    monthlyFee: this.monthlyFee,
    currency: this.currency,
    totalHouses: this.totalHouses,
    isActive: this.isActive,
    createdAt: this.createdAt,
  };
};

const Community = mongoose.model("Community", communitySchema);

module.exports = Community;

File: app/lib/models/community_model.dart
class Community {
  final String id;
  final String name;
  final String address;
  final String description;
  final double monthlyFee; // Mensualidad base
  final String currency; // Moneda (MXN, USD, etc.)
  final int totalHouses; // Total de casas/habitaciones
  final DateTime createdAt;
  final DateTime? updatedAt;
  final bool isActive;
  final Map<String, dynamic> settings; // Configuraciones específicas
  final List<String> adminIds; // IDs de administradores
  final String superAdminId; // ID del super administrador

  Community({
    required this.id,
    required this.name,
    required this.address,
    required this.description,
    required this.monthlyFee,
    required this.currency,
    required this.totalHouses,
    required this.createdAt,
    this.updatedAt,
    required this.isActive,
    required this.settings,
    required this.adminIds,
    required this.superAdminId,
  });

  factory Community.fromJson(Map<String, dynamic> json) {
    return Community(
      id: json['_id'] ?? json['id'] ?? '', // MongoDB usa _id, pero también acepta id
      name: json['name'] ?? '',
      address: json['address'] ?? '',
      description: json['description'] ?? '',
      monthlyFee: (json['monthlyFee'] ?? 0.0).toDouble(),
      currency: json['currency'] ?? 'MXN',
      totalHouses: json['totalHouses'] ?? 0,
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt'].toString()) : DateTime.now(),
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt'].toString()) : null,
      isActive: json['isActive'] ?? true,
      settings: json['settings'] is Map<String, dynamic> ? json['settings'] : {},
      adminIds: json['adminIds'] is List ? List<String>.from(json['adminIds'].map((id) => id.toString())) : [],
      superAdminId: json['superAdminId']?["_id"]?.toString() ?? json['superAdminId']?.toString() ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'address': address,
      'description': description,
      'monthlyFee': monthlyFee,
      'currency': currency,
      'totalHouses': totalHouses,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
      'isActive': isActive,
      'settings': settings,
      'adminIds': adminIds,
      'superAdminId': superAdminId,
    };
  }

  Community copyWith({
    String? id,
    String? name,
    String? address,
    String? description,
    double? monthlyFee,
    String? currency,
    int? totalHouses,
    DateTime? createdAt,
    DateTime? updatedAt,
    bool? isActive,
    Map<String, dynamic>? settings,
    List<String>? adminIds,
    String? superAdminId,
  }) {
    return Community(
      id: id ?? this.id,
      name: name ?? this.name,
      address: address ?? this.address,
      description: description ?? this.description,
      monthlyFee: monthlyFee ?? this.monthlyFee,
      currency: currency ?? this.currency,
      totalHouses: totalHouses ?? this.totalHouses,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      isActive: isActive ?? this.isActive,
      settings: settings ?? this.settings,
      adminIds: adminIds ?? this.adminIds,
      superAdminId: superAdminId ?? this.superAdminId,
    );
  }

  @override
  String toString() {
    return 'Community(id: $id, name: $name, address: $address, totalHouses: $totalHouses, isActive: $isActive)';
  }
}
